### PR TITLE
New checksum APIs in v1model; better alias analysis in side-effects; …

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
…removed some hacks from bmv2 backend which were related to checksums

This patch still does not generate JSON code for verifying checksum, but it should be very easy to add that capability now. There is a whole set of issues that should be addressed by this PR; I will list them separately.

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
This PR should fix the following issues: #134, #445, #609, #741, #775, #840

**Comment by mbudiu-vmw** [comment]
This PR simplifies the way checksums are computed and verified in v1model by using an extern function for each API; no extern objects exist anymore. This is much closer to the way BMv2 (and P4-14) handle checksums, so translation is much simpler in both directions. This new API also makes it much easier to give good error messages when users misuse the verify/update checksum controls.

The better alias analysis was needed so that the compiler does not mess up these nice extern function calls. On the positive side, a bunch of relatively hacky code to handle checksums in the BMv2 back-end has been completely removed.

File: frontends/p4/sideEffects.cpp [hunk 1]
```diff
@@ -310,40 +310,206 @@ class DismantleExpression : public Transform {
...
-    // We don't want to compute the full read/write set here so we
-    // overapproximate it as follows: all declarations that occur in
-    // an expression.
-    // TODO: this could be made more precise, perhaps using LocationSets.
-    class ReadsWrites : public Inspector {
+    /// This class represents the path to a location.
+    /// Given a struct S { bit a; bit b; } and a variable S x;
+    /// a path can be x.a, or just x.  An array index is represented as a
+    /// number (encoded as a string) or as "*", denoting an unknown index.
+    struct LocationPath : public IHasDbPrint {
```
Comment (antoninbas) [inline frontends/p4/sideEffects.cpp]: Would have been nice to make the changes to this pass in a separate PR

File: p4include/v1model.p4 [hunk 2]
```diff
@@ -140,6 +135,27 @@ extern void clone3<T>(in CloneType type, in bit<32> session, in T data);
...
+Verifies the 16-bit checksum of the supplied data.
+T must be a list expression where all the fields are bit-fields or varbits
+and where the total dynamic length of the fields is a multiple of 16 bits.
+If this method detects that a checksum of the data is not correct it
+sets an internal error flag.
+@param condition  If 'false' the verification always succeeds.
+@param data       Data whose checksum is verified.
+@param checksum   Expected checksum of the data.
+*/
+extern void verify_checksum<T>(in bool condition, in T data, in bit<16> checksum);
```
Comment (antoninbas) [inline p4include/v1model.p4]: is there a reason to limit these new extern methods to the "csum16" algorithm?
why not support `verify_checksum<T, U>(in bool condition, in T data, in U checksum, in HashAlgorithm algo)`?

File: p4include/v1model.p4 [hunk 3]
```diff
@@ -140,6 +135,27 @@ extern void clone3<T>(in CloneType type, in bit<32> session, in T data);
 
 extern void truncate(in bit<32> length);
 
+/**
+Verifies the 16-bit checksum of the supplied data.
+T must be a list expression where all the fields are bit-fields or varbits
+and where the total dynamic length of the fields is a multiple of 16 bits.
+If this method detects that a checksum of the data is not correct it
+sets an internal error flag.
```
Comment (antoninbas) [inline p4include/v1model.p4]: isn't the plan to expose this information as standard_metadata which can be consumed in ingress?

File: frontends/p4/sideEffects.cpp (hunk 1 above)
Comment (mbudiu-vmw) [inline frontends/p4/sideEffects.cpp]: yes, but this PR would depend on that other one for being merged first.

File: p4include/v1model.p4 (hunk 3 above)
Comment (mbudiu-vmw) [inline p4include/v1model.p4]: Yes, but that will be a separate PR.

File: p4include/v1model.p4 (hunk 2 above)
Comment (mbudiu-vmw) [inline p4include/v1model.p4]: No, but currently this is the only algorithm supported - at least in the back-end, so I didn't want to make the API more complicated than necessary.

Comment (antoninbas) [inline p4include/v1model.p4]: I think that was my question. Is there a reason not to support the other algorithms in the backend? It looks like you would just need to map the enum value to the algo name and call `createCalculation` with that name as the first argument. It may be worth it to add support now rather than modify v1model.p4 in a non-backward compatible fashion in the future.

Comment (mbudiu-vmw) [inline p4include/v1model.p4]: I will amend the PR with support for an algorithm parameter.
Only csum16 will be supported initially.

**Comment by mbudiu-vmw** [comment]
@antoninbas : this additional commit adds the Algorithm argument to the two checksum methods.

**Comment by antoninbas** [comment]
@mbudiu-vmw I believe that the destination field's type need to be a generic type (e.g. needs to be `bit<32>` for a crc32 algorithm).

**Comment by mbudiu-vmw** [comment]
I hope I have addressed all of @antoninbas comments.
Still pending: doing checksum verification - issue #675 - and separating verification from update.
I will probably have a single separate PR for both of these.
